### PR TITLE
feat: update token.sol to track volume

### PR DESCRIPTION
**Pull request by glo12345**

### CODE REVIEW

File: contracts/ethereum/contracts/LootboxStaking.sol [hunk 1]
```diff
@@ -33,19 +33,19 @@ import "./openzeppelin/SafeERC20.sol";
...
-uint256 constant MIN_LOCKUP_TIME = 31 days;
-
-uint256 constant MAX_LOCKUP_TIME = 365 days;
-
 uint256 constant MAX_SLIPPAGE = 100;
 
 contract LootboxStaking is ILootboxStaking, IOperatorOwned, IEmergencyMode {
     using SafeERC20 for IERC20;
 
+    uint256 private minLockupTime_;
```
Comment (af-afk) [inline contracts/ethereum/contracts/LootboxStaking.sol:45]: These changes have to be at the bottom of the states due to the upgrade compatibility, and you can't change the constants instead?

File: contracts/ethereum/contracts/LootboxStaking.sol [hunk 2]
```diff
@@ -75,7 +75,7 @@ contract LootboxStaking is ILootboxStaking, IOperatorOwned, IEmergencyMode {
 
     ICamelotPair private camelotFusdcWethPair_;
 
-    mapping (address => Deposit[]) private deposits_;
+    mapping(address => Deposit[]) private deposits_;
```
Comment (af-afk) [inline contracts/ethereum/contracts/LootboxStaking.sol:78]: Same issue with the other

File: contracts/ethereum/contracts/LootboxStaking.sol [hunk 3]
```diff
@@ -99,6 +99,8 @@ contract LootboxStaking is ILootboxStaking, IOperatorOwned, IEmergencyMode {
     function init(
         address _operator,
         address _emergencyCouncil,
+        uint256 _minLockupTime,
```
Comment (af-afk) [inline contracts/ethereum/contracts/LootboxStaking.sol:102]: Need a separate function to initialise these variables and need to bump up the version

File: contracts/ethereum/contracts/Token.sol [hunk 4]
```diff
@@ -165,21 +165,33 @@ contract Token is
     /// @notice burnFee_ that's paid by the user when they mint
     uint256 private mintFee_;
 
+    /* ~~~~~~~~~~ VOLUME TRACKNG ~~~~~~~~~~ */
+
+    struct EpochVolume {
```
Comment (af-afk) [inline contracts/ethereum/contracts/Token.sol]: Can you put the struct outside the scope of this and into an interface? And add documentation explaining what it does? And can you not use a mapping inside the struct?

File: contracts/ethereum/contracts/Token.sol [hunk 5]
```diff
@@ -401,14 +406,26 @@ contract Token is
...
+        );
 
         unchecked {
             balances_[from] = fromBalance - amount;
         }
 
         balances_[to] += amount;
 
+        // Get epoch block range
+        // TODO
```
Comment (af-afk) [inline contracts/ethereum/contracts/Token.sol:421]: Don't leave the comment here

File: contracts/ethereum/contracts/Token.sol [hunk 6]
```diff
@@ -418,6 +435,16 @@ contract Token is
 
         totalSupply_ += _amount;
         balances_[_account] += _amount;
+
+        // Get epoch block range
+        // TODO
+        uint256 startEpoch = 0;
+
+        // set epoch volume
```
Comment (af-afk) [inline contracts/ethereum/contracts/Token.sol:442]: Maybe you can eliminate some redundancy here

File: contracts/ethereum/contracts/Token.sol [hunk 7]
```diff
@@ -429,16 +456,26 @@ contract Token is
...
         unchecked {
             balances_[_account] = accountBalance - _amount;
-
         }
 
         totalSupply_ -= _amount;
 
+        // Get epoch block range
+        // TODO
+        uint256 startEpoch = 0;
```
Comment (af-afk) [inline contracts/ethereum/contracts/Token.sol:469]: ... same

File: contracts/ethereum/contracts/Token.sol [hunk 8]
```diff
@@ -869,4 +928,17 @@ contract Token is
         mintFee_ = _mintFee;
         burnFee_ = _burnFee;
     }
+
+    /* ~~~~~~~~~~ EPOCH VOLUME FUNCTIONS ~~~~~~~~~~ */
+
+    function volume(
```
Comment (af-afk) [inline contracts/ethereum/contracts/Token.sol]: Can you make this a part of the interface too?

**Review by af-afk — CHANGES_REQUESTED** [review]

**Comment by glo12345** [comment]
Ditching tracking on-chain to track utility specific volume - migrating to develop-add-trading-comp-leaderboard

**Pull request closed without merging**